File: src/core/kernel/event-bus.ts
/**
 * EventBus — 小奕 J.A.R.V.I.S. 事件总线
 * @core/kernel
 *
 * 功能：
 * 1. 模块间解耦通信
 * 2. 事件订阅/发布/取消
 * 3. 异步事件处理
 * 4. 事件历史记录
 * 5. 错误隔离（单个处理器失败不影响其他）
 *
 * 萃取自 EventEmitter2 架构设计，重写为小奕原生实现
 */

// ============================================================
// 类型定义
// ============================================================

export interface EventBusEvent<T = unknown> {
  type: string;
  payload: T;
  timestamp: number;
  source?: string;
  correlationId?: string;
}

export type EventHandler<T = unknown> = (event: EventBusEvent<T>) => void | Promise<void>;

export interface EventSubscription {
  id: string;
  eventType: string;
  handler: EventHandler;
  once: boolean;
  active: boolean;
}

export interface EventBusConfig {
  maxHistory: number;
  enableAsync: boolean;
  errorStrategy: 'log' | 'ignore' | 'propagate';
}

// ============================================================
// EventBus 实现
// ============================================================

export class EventBus {
  private subscriptions: Map<string, EventSubscription[]> = new Map();
  private history: EventBusEvent[] = [];
  private config: EventBusConfig;
  private subscriptionIdCounter = 0;

  constructor(config: Partial<EventBusConfig> = {}) {
    this.config = {
      maxHistory: config.maxHistory ?? 1000,
      enableAsync: config.enableAsync ?? true,
      errorStrategy: config.errorStrategy ?? 'log',
    };
  }

  // ============================================================
  // 订阅管理
  // ============================================================

  /**
   * 订阅事件
   * @param eventType 事件类型（支持通配符 '*'）
   * @param handler 事件处理器
   * @param once 是否仅触发一次
   * @returns 订阅 ID（可用于取消订阅）
   */
  subscribe<T = unknown>(
    eventType: string,
    handler: EventHandler<T>,
    once: boolean = false,
  ): string {
    const subscriptionId = `sub_${++this.subscriptionIdCounter}`;

    const subscription: EventSubscription = {
      id: subscriptionId,
      eventType,
      handler: handler as EventHandler,
      once,
      active: true,
    };

    const existing = this.subscriptions.get(eventType) || [];
    existing.push(subscription);
    this.subscriptions.set(eventType, existing);

    return subscriptionId;
  }

  /**
   * 取消订阅
   */
  unsubscribe(subscriptionId: string): boolean {
    for (const [eventType, subs] of this.subscriptions.entries()) {
      const index = subs.findIndex(s => s.id === subscriptionId);
      if (index !== -1) {
        subs[index].active = false;
        subs.splice(index, 1);
        return true;
      }
    }
    return false;
  }

  /**
   * 取消某个事件类型的所有订阅
   */
  unsubscribeAll(eventType: string): number {
    const subs = this.subscriptions.get(eventType);
    if (!subs) return 0;
    const count = subs.length;
    subs.forEach(s => (s.active = false));
    this.subscriptions.delete(eventType);
    return count;
  }

  // ============================================================
  // 事件发布
  // ============================================================

  /**
   * 发布事件（同步）
   */
  emit<T = unknown>(eventType: string, payload: T, source?: string): void {
    const event: EventBusEvent<T> = {
      type: eventType,
      payload,
      timestamp: Date.now(),
      source,
    };

    // 记录历史
    this._recordHistory(event);

    // 触发处理器
    this._dispatch(event);
  }

  /**
   * 发布事件（异步）
   */
  async emitAsync<T = unknown>(
    eventType: string,
    payload: T,
    source?: string,
  ): Promise<void> {
    const event: EventBusEvent<T> = {
      type: eventType,
      payload,
      timestamp: Date.now(),
      source,
    };

    this._recordHistory(event);
    await this._dispatchAsync(event);
  }

  // ============================================================
  // 内部方法
  // ============================================================

  private _dispatch(event: EventBusEvent): void {
    // 精确匹配
    const handlers = this.subscriptions.get(event.type) || [];

    // 通配符匹配
    const wildcardHandlers = this.subscriptions.get('*') || [];

    const allHandlers = [...handlers, ...wildcardHandlers];

    for (const sub of allHandlers) {
      if (!sub.active) continue;

      try {
        sub.handler(event);

        if (sub.once) {
          sub.active = false;
        }
      } catch (error) {
        this._handleError(error, sub, event);
      }
    }
  }

  private async _dispatchAsync(event: EventBusEvent): Promise<void> {
    const handlers = this.subscriptions.get(event.type) || [];
    const wildcardHandlers = this.subscriptions.get('*') || [];
    const allHandlers = [...handlers, ...wildcardHandlers];

    const promises = allHandlers
      .filter(sub => sub.active)
      .map(async sub => {
        try {
          const result = sub.handler(event);
          if (result instanceof Promise) {
            await result;
          }

          if (sub.once) {
            sub.active = false;
          }
        } catch (error) {
          this._handleError(error, sub, event);
        }
      });

    await Promise.allSettled(promises);
  }

  private _handleError(error: unknown, sub: EventSubscription, event: EventBusEvent): void {
    const errorMessage = error instanceof Error ? error.message : String(error);

    switch (this.config.errorStrategy) {
      case 'log':
        console.error(`[EventBus] 处理器错误 (${sub.id}):`, errorMessage, {
          eventType: event.type,
          handler: sub.handler,
        });
        break;
      case 'ignore':
        // 静默忽略
        break;
      case 'propagate':
        throw error;
    }
  }

  private _recordHistory(event: EventBusEvent): void {
    this.history.push(event);

    // 限制历史大小
    while (this.history.length > this.config.maxHistory) {
      this.history.shift();
    }
  }

  // ============================================================
  // 查询方法
  // ============================================================

  /**
   * 获取事件历史
   */
  getHistory(eventType?: string, limit: number = 100): EventBusEvent[] {
    let events = this.history;

    if (eventType) {
      events = events.filter(e => e.type === eventType || eventType === '*');
    }

    return events.slice(-limit);
  }

  /**
   * 获取订阅数量
   */
  getSubscriptionCount(eventType?: string): number {
    if (eventType) {
      return (this.subscriptions.get(eventType) || []).filter(s => s.active).length;
    }

    let count = 0;
    for (const subs of this.subscriptions.values()) {
      count += subs.filter(s => s.active).length;
    }
    return count;
  }

  /**
   * 清除历史
   */
  clearHistory(): void {
    this.history = [];
  }

  /**
   * 销毁事件总线
   */
  destroy(): void {
    this.subscriptions.clear();
    this.history = [];
    this.subscriptionIdCounter = 0;
  }
}

// ============================================================
// 单例导出
// ============================================================

export const globalEventBus = new EventBus();

File: src/core/kernel/event_bus.py
"""
Event Bus module for J.A.R.V.I.S.
Provides pub/sub event system with history tracking.
"""

import threading
import time
import uuid
from collections import deque
from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Callable, Dict, List, Optional


class EventType(Enum):
    """Standard event types for J.A.R.V.I.S. system events."""
    SYSTEM = "system"
    USER = "user"
    AGENT = "agent"
    PLUGIN = "plugin"
    ERROR = "error"


@dataclass
class Event:
    """Event data container."""
    event_type: str
    source: str
    data: Any
    timestamp: str = field(default_factory=lambda: __import__('time').strftime("%Y-%m-%dT%H:%M:%S"))
    correlation_id: str = field(default_factory=lambda: str(uuid.uuid4()))

    @property
    def payload(self) -> Any:
        """Alias for data."""
        return self.data

    @property
    def type(self) -> str:
        """Alias for event_type."""
        return self.event_type


class EventBus:
    """Thread-safe event bus with pub/sub and history tracking."""

    def __init__(self, max_history: int = 200):
        self._history: deque = deque(maxlen=max_history)
        self._subscribers: Dict[str, List[Dict]] = {}
        self._sub_id_counter = 0
        self._lock = threading.Lock()

    def subscribe(self, event_type: str, callback: Callable, once: bool = False) -> int:
        """Subscribe to an event type. Returns subscription ID."""
        with self._lock:
            sub_id = self._sub_id_counter
            self._sub_id_counter += 1
            self._subscribers.setdefault(event_type, []).append(
                {"id": sub_id, "callback": callback, "once": once}
            )
        return sub_id

    def unsubscribe(self, sub_id: int) -> bool:
        """Unsubscribe by ID. Returns True if found."""
        if type(sub_id) is not int:
            return False
        removed = False
        with self._lock:
            for event_type in list(self._subscribers):
                subs = self._subscribers[event_type]
                self._subscribers[event_type] = [s for s in subs if s["id"] != sub_id]
                if len(self._subscribers[event_type]) != len(subs):
                    removed = True
                if not self._subscribers[event_type]:
                    del self._subscribers[event_type]
        return removed

    def emit(self, event_type: str, payload: Any = None) -> None:
        """Emit an event to all subscribers."""
        event = Event(event_type=event_type, source="event_bus", data=payload)
        self._dispatch(event)

    def _dispatch(self, event: Event) -> None:
        """Record and dispatch the supplied event without changing its source."""
        with self._lock:
            self._history.append(event)

        self._notify(event)

    def _notify(self, event: Event) -> None:
        """Notify subscribers for an already recorded event."""
        subscribers_to_call = []
        with self._lock:
            subscription_types = [event.event_type]
            if event.event_type != "*":
                subscription_types.append("*")
            for subscription_type in subscription_types:
                subscribers = self._subscribers.get(subscription_type)
                if not subscribers:
                    continue
                subscribers_to_call.extend(subscribers)
                retained = [sub for sub in subscribers if not sub["once"]]
                if retained:
                    self._subscribers[subscription_type] = retained
                else:
                    del self._subscribers[subscription_type]

        for sub in subscribers_to_call:
            try:
                sub["callback"](event)
            except Exception:
                pass

    def publish(self, event: Event) -> None:
        """Publish an Event object."""
        self._dispatch(event)

    def record_many(
        self, events: List[Event], *, deadline: float | None = None
    ) -> bool:
        """Atomically record events without invoking subscribers."""
        pending = list(events)
        if any(not isinstance(event, Event) for event in pending):
            raise ValueError("events must contain only Event objects")
        if deadline is None:
            self._lock.acquire()
        else:
            remaining = deadline - time.monotonic()
            if remaining <= 0 or not self._lock.acquire(timeout=remaining):
                return False
            if time.monotonic() >= deadline:
                self._lock.release()
                return False
        try:
            self._history.extend(pending)
            return True
        finally:
            self._lock.release()

    def get_history(self, event_type: Optional[str] = None, limit: int = 100) -> List[Event]:
        """Get event history, optionally filtered by type."""
        if type(limit) is not int or limit < 0:
            raise ValueError("limit must be a non-negative integer")
        if limit == 0:
            return []
        with self._lock:
            history = list(self._history)
        if event_type is not None:
            history = [e for e in history if e.event_type == event_type]
        return history[-limit:]

    def clear_history(self) -> None:
        """Clear all event history."""
        with self._lock:
            self._history.clear()

    def clear(self) -> None:
        """Clear event history."""
        self.clear_history()

    def get_subscription_count(self) -> int:
        """Get total active subscriptions."""
        with self._lock:
            return sum(len(subs) for subs in self._subscribers.values())

    def destroy(self) -> None:
        """Destroy: clear all subscriptions and history."""
        with self._lock:
            self._subscribers.clear()
            self._history.clear()
